### PR TITLE
Load and plot uploads

### PR DESCRIPTION
**Pull request by adelavega**
Adds `load_uploads` and `plot_uploads` functions to the `Analysis` class, which facilitate downloading statmap images directly from Neurovault collections, and plotting them using nilearn.

`load_uploads` allows you to filter the attributes by which to filter collections you want to load (i.e. only load collections with `estimator` = `afni`), filter the images themselves (i.e. only group images), and loads the resulting imagines in a flat list of tuples across collections, with the first item being a `nibabel.NiftiImage` object. Images are downloaded into a directory which is user specified or a temp directory automatically created.

For example, given a `Analysis` object `m`:

```
m.get_uploads()
[{'cli_version': '0.4.1',
  'collection_id': 9370,
  'estimator': 'nistats',
  'files': [{'basename': 'task-movie_space-MNI152NLin2009cAsym_contrast-building_stat-t_statmap.nii.gz',
    'level': 'GROUP',
    'status': 'OK',
    'traceback': None},
   {'basename': 'task-movie_space-MNI152NLin2009cAsym_contrast-building_stat-effect_statmap.nii.gz',
    'level': 'GROUP',
    'status': 'OK',
    'traceback': None},
   {'basename': 'task-movie_space-MNI152NLin2009cAsym_contrast-building_stat-variance_statmap.nii.gz',
    'level': 'GROUP',
    'status': 'OK',
    'traceback': None}],
  'fmriprep_version': '20.1.1',
  'uploaded_at': '2021-01-22T11:2'},
 {'cli_version': None,
  'collection_id': 8760,
  'estimator': None,
  'files': [{'basename': 'task-movie_space-MNI152NLin2009cAsym_contrast-building_stat-variance_statmap.nii.gz',
    'level': 'GROUP',
    'status': 'OK',
    'traceback': None},
   {'basename': 'task-movie_space-MNI152NLin2009cAsym_contrast-building_stat-t_statmap.nii.gz',
    'level': 'GROUP',
    'status': 'OK',
    'traceback': None},
   {'basename': 'task-movie_space-MNI152NLin2009cAsym_contrast-building_stat-effect_statmap.nii.gz',
    'level': 'GROUP',
    'status': 'OK',
    'traceback': None}],
  'fmriprep_version': None,
  'uploaded_at': '2020-09-28T18:1'},
 {'cli_version': '0.4.1',
  'collection_id': 9383,
  'estimator': 'afni',
  'files': [{'basename': 'task-movie_space-MNI152NLin2009cAsym_contrast-building_stat-t_statmap.nii.gz',
    'level': 'GROUP',
    'status': 'OK',
    'traceback': None},
   {'basename': 'task-movie_space-MNI152NLin2009cAsym_contrast-building_stat-effect_statmap.nii.gz',
    'level': 'GROUP',
    'status': 'OK',
    'traceback': None},
   {'basename': 'task-movie_space-MNI152NLin2009cAsym_contrast-building_stat-variance_statmap.nii.gz',
    'level': 'GROUP',
    'status': 'OK',
    'traceback': None}],
  'fmriprep_version': '20.1.1',
  'uploaded_at': '2021-01-23T02:1'}]
```

In this scenario there are three uploaded collections, with three images each.

By default, only images from the latest collection are loaded:

```
m.load_uploads()

[(<nibabel.nifti1.Nifti1Image at 0x7fc493ad7b50>,
  {'cli_version': '0.4.1',
   'collection_id': 9383,
   'estimator': 'afni',
   'fmriprep_version': '20.1.1',
   'uploaded_at': datetime.datetime(2021, 1, 23, 2, 1),
   'basename': 'task-movie_space-MNI152NLin2009cAsym_contrast-building_stat-t_statmap.nii.gz',
   'level': 'GROUP'}),
 (<nibabel.nifti1.Nifti1Image at 0x7fc493b4b5d0>,
  {'cli_version': '0.4.1',
   'collection_id': 9383,
   'estimator': 'afni',
   'fmriprep_version': '20.1.1',
   'uploaded_at': datetime.datetime(2021, 1, 23, 2, 1),
   'basename': 'task-movie_space-MNI152NLin2009cAsym_contrast-building_stat-effect_statmap.nii.gz',
   'level': 'GROUP'}),
 (<nibabel.nifti1.Nifti1Image at 0x7fc494b76850>,
  {'cli_version': '0.4.1',
   'collection_id': 9383,
   'estimator': 'afni',
   'fmriprep_version': '20.1.1',
   'uploaded_at': datetime.datetime(2021, 1, 23, 2, 1),
   'basename': 'task-movie_space-MNI152NLin2009cAsym_contrast-building_stat-variance_statmap.nii.gz',
   'level': 'GROUP'})]
```

But you can also load all:
```
m.load_uploads(select=None)

[(<nibabel.nifti1.Nifti1Image at 0x7fc493ad7e10>,
  {'cli_version': None,
   'collection_id': 8760,
   'estimator': None,
   'fmriprep_version': None,
   'uploaded_at': datetime.datetime(2020, 9, 28, 18, 1),
   'basename': 'task-movie_space-MNI152NLin2009cAsym_contrast-building_stat-variance_statmap.nii.gz',
   'level': 'GROUP'}),
 (<nibabel.nifti1.Nifti1Image at 0x7fc493ad7650>,
  {'cli_version': None,
   'collection_id': 8760,
   'estimator': None,
   'fmriprep_version': None,
   'uploaded_at': datetime.datetime(2020, 9, 28, 18, 1),
   'basename': 'task-movie_space-MNI152NLin2009cAsym_contrast-building_stat-t_statmap.nii.gz',
   'level': 'GROUP'}),
 (<nibabel.nifti1.Nifti1Image at 0x7fc493ae3b50>,
  {'cli_version': None,
   'collection_id': 8760,
   'estimator': None,
   'fmriprep_version': None,
   'uploaded_at': datetime.datetime(2020, 9, 28, 18, 1),
   'basename': 'task-movie_space-MNI152NLin2009cAsym_contrast-building_stat-effect_statmap.nii.gz',
   'level': 'GROUP'}),
 (<nibabel.nifti1.Nifti1Image at 0x7fc493afbbd0>,
  {'cli_version': '0.4.1',
   'collection_id': 9370,
   'estimator': 'nistats',
   'fmriprep_version': '20.1.1',
   'uploaded_at': datetime.datetime(2021, 1, 22, 11, 2),
   'basename': 'task-movie_space-MNI152NLin2009cAsym_contrast-building_stat-t_statmap.nii.gz',
   'level': 'GROUP'}),
 (<nibabel.nifti1.Nifti1Image at 0x7fc493afbd90>,
  {'cli_version': '0.4.1',
   'collection_id': 9370,
   'estimator': 'nistats',
   'fmriprep_version': '20.1.1',
   'uploaded_at': datetime.datetime(2021, 1, 22, 11, 2),
   'basename': 'task-movie_space-MNI152NLin2009cAsym_contrast-building_stat-effect_statmap.nii.gz',
   'level': 'GROUP'}),
 (<nibabel.nifti1.Nifti1Image at 0x7fc493afdad0>,
  {'cli_version': '0.4.1',
   'collection_id': 9370,
   'estimator': 'nistats',
   'fmriprep_version': '20.1.1',
   'uploaded_at': datetime.datetime(2021, 1, 22, 11, 2),
   'basename': 'task-movie_space-MNI152NLin2009cAsym_contrast-building_stat-variance_statmap.nii.gz',
   'level': 'GROUP'}),
 (<nibabel.nifti1.Nifti1Image at 0x7fc493afd690>,

...
```

`select` can also be `oldest` or `newest` (explicit).

You could also pass a dictionary `coll_filt` to filter the collections. 
e.g.:
```
m.load_uploads(coll_filt={'estimator': 'afni})
```

kwargs can be used to filter the images themselves:

```
m.load_uploads(coll_filt={'estimator': 'afni}, 'level'='GROUP')
```

Finally, `plot_uploads` uses `plot_stat_map` to plot these images:

```
m.plot_uploads()
```
image
image
image

will plot the images from the latest collection. `kwargs` go to `load_uploads` so it can be specified which images to plot. 

@rbroc can you take a look at this and give it a go? 
just want to see if this is useful and it makes sense how I implemented this.

Things still left to do:
- Extract BIDS entities from the file names of images, to allow filtering based on that (i.e. only plot `t` maps, or filter by contrast name
- Maybe make the default plotting parameters better, and add overlays indicating which images is being plotting (you can use `plot_args` to mass arguments to `plot_stat_map`).
Thanks!

### CODE REVIEW

**Comment by rbroc** [comment]
definitely super useful functionality IMO - I've tried on some of the neuroscout-paper analyses and if no filters are passed the loader function works like a charm.
Can't find analyses with multiple uploads to test the `select` argument (wondering what `sorted` does if `select` is `None`?)
Two more remarks/suggestions: 
1. Thinking of a scenario where users wants to iteratively load images for more than one analysis (e.g. meta-analysis contexts) where this functionality comes in very handy, could it be useful to add to metadata (contrast, level, dataset) to the output?
2. I can't get the `coll_filter` and `stat_filter` to work. Haven't tried super thoroughly, but how they work is not intuitive to me - possibly because I am not super familiar with NV. Would it be an option to allow users to filter in a way that's closer to Neuroscout's ontology (e.g. by contrast or level), and maybe with regex instead of exact match? 
3. If the former is a bad idea, can the way filtering works be made more straightforward for the user (e.g. in the docstring)?
4. Plotting fails for me for the second item returned for `M4jEG` with this error: `ValueError: Image size of 539x274584 pixels is too large. It must be less than 2^16 in each direction.` Also not so sure I'd spend too much time on plotting, it could be left to the user and how nice things look always depends on the specific use case. More something for a `neuroscout-paper` repo-like set of utils?
Hope that makes sense!

File: pyns/models/analysis.py [hunk 1]
```diff
@@ -378,6 +386,71 @@ def get_uploads(self, id):
...
+        uploads = sorted(uploads, key=lambda x: x['uploaded_at'], reverse=(select=='latest'))
+
+        # Select collections based on filters
+        uploads = [
+            u for u in uploads 
+            if all([u.get(k, None) == v for k, v in coll_filt.items()])
+            ] 
+
+        if select is not None:
+            uploads = [uploads[0]]
```
Comment (rbroc) [inline pyns/models/analysis.py:415]: This fails if uploads is empty - maybe add explicit error for this scenario?

**Review by rbroc — COMMENTED** [review]
left some general comments in the thread and very minor thing in the code

**Comment by adelavega** [comment]
1) yes I'm going to add at least the meta-data in the file name. if we need anything else specifically after that we can add on a case by case basis
2-3) I will try to clarify it in the docstring. i'm actually just using the information from what you would get if you call `get_uploads`. those are the fields you can filter on. not accessing any neurovault specific information actually. 
do you have example of something you tried?
4) weird that sounds like nilearn thing. I agree not going to spend too much time on it, maybe just add a formatted title to make it easier to see which is plotted. 

awesome, thanks. the only real concern i have still is if filtering is not working, so let me know if you have that example.

**Comment by adelavega** [comment]
Okay @rbroc I simplified it so you can simply combine all filter arguments such as:

    m.load_uploads(estimator='nistats', level='GROUP')

This makes the API a bit cleaner, although it mixes the two levels. I don't forsee any clashes in the key names though.

**Pull request merged**